sys/net/application_layer/sock_dns/dns_cache.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (C) 2022 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup net_sock_dns
+ * @{
+ * @file
+ * @brief   DNS cache implementation
+ * @author  Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ * @}
+ */
+
+#include "checksum/fletcher32.h"
+#include "net/sock/dns.h"
+#include "dns_cache.h"
+#include "ztimer.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+static struct dns_cache_entry {
+    uint32_t hash;
+    uint32_t expires;
+    union {
+#if IS_ACTIVE(SOCK_HAS_IPV4)
+        ipv4_addr_t v4;
+#endif
+#if IS_ACTIVE(SOCK_HAS_IPV6)
+        ipv6_addr_t v6;
+#endif
+    } addr;
+    uint8_t len;
+} cache[CONFIG_DNS_CACHE_SIZE];
+
+static uint8_t _addr_len(int family)
+{
+    switch (family) {
+    case AF_INET:
+        return 4;
+    case AF_INET6:
+        return 16;
+    }
+    return 0;
+}
+
+int sock_dns_cache_query(const char *domain_name, void *addr_out, int family)
+{
+    uint32_t now = ztimer_now(ZTIMER_MSEC) / MS_PER_SEC;
+    uint32_t hash = fletcher32((void *)domain_name, strlen(domain_name) / 2);
+    uint8_t addr_len = _addr_len(family);
+
+    for (unsigned i = 0; i < CONFIG_DNS_CACHE_SIZE; ++i) {
+        /* empty slot */
+        if (cache[i].len == 0) {
+            continue;
+        }
```

Comment:
Would it be possible to cache queries with a family of `AF_UNSPEC´ at the moment? As `_addr_len` returns 0 in this case, this doesn't seem to be the case. Or is this irrelevant as the family is always either `AF_INET` or `AF_INET6`?

Reply:
The response to an `AF_UNSPEC` query can't be unspecified, it will have a length.

Reply:
Oh, I see. Because of the `!addr_len` in the check this also doesn't affect caching, right?